webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -62,14 +62,14 @@ object ReportingServiceUtils {
    * Build rule status reports from node reports, deciding which directives should be "skipped"
    */
   def buildRuleStatusReport(ruleId: RuleId, nodeReports: Map[NodeId, NodeStatusReport]): RuleStatusReport = {
-    val toKeep     = nodeReports.values.flatMap(_.reports.flatMap(_._2.reports)).filter(_.ruleId == ruleId).toList
+    val toKeep     = nodeReports.values.flatMap(_.reports.flatMap(_._2.reports)).filter(_.ruleId == ruleId)
     // we don't keep overrides for a directive which is already in "toKeep" or that don't target that rule
-    val toKeepDir  = toKeep.map(_.directives.keySet).toSet.flatten
+    val toKeepDir  = toKeep.flatMap(_.directives.keySet).toSet
```

Comment:
massive optimisazation 2

Reply:
i was suprised that it needed a toSet to stay a set

Reply:
not the same set type